File: src/lambda/textnow/send_message.js
const shlex = require('shlex');
const textnow = require('textnow-bot');
const { replyMessage, replyPhoto } = require('../../common/actions');
const { SCREENSHOT_PATH, execute } = require('../../common/puppeteer');

module.exports.handler = async event => {
  const interface = event.interface;
  const user = event.dynamodb.Item;
  let args = null;

  try {
    args = shlex.split(event.command).slice(1);
  }
  catch (error) {
    await replyMessage(interface, 'Failed to parse input.');
    throw error;
  }

  if (!user.cookies) {
    await replyMessage(interface, 'Please log into TextNow first.');
    throw new Error('Missing login cookies');
  }

  if (args.length < 2) {
    await replyMessage(interface, 'Please provide a recipient and message.');
    throw new Error('Missing recipient / message info');
  }

  await replyMessage(interface, 'Sending message...');

  try {
    const [recipient, ...messageArgs] = args;
    const message = messageArgs.join(' ');

    const bot = async (page, client) => {
      await textnow.logIn(page, client);
      await textnow.selectConversation(page, recipient);
      await textnow.sendMessage(page, message);

      const cookies = (await client.send('Network.getAllCookies')).cookies;
      return cookies;
    };

    user.cookies = await execute(bot, user.cookies);
  }
  catch (error) {
    await replyPhoto(interface, SCREENSHOT_PATH);
    await replyMessage(interface, 'Failed to send message - input may be invalid.');
    throw error;
  }

  await replyMessage(interface, 'Successfully sent message!');
  return user;
};

File: lib/textnow/lambda_function/send_message.py
import shlex
import textnow_bot

from lib.core.dynamodb.users_table import UsersTable
from lib.core.interface import Interface
from lib.core.playwright import Playwright

def handler(event, _):
  interface = Interface(event['interface'])
  users_table = UsersTable(event['userId'])

  user = users_table.get_user()

  try:
    args = shlex.split(event['command'])[1:]
  except:
    interface.reply_message('Failed to parse input.')
    raise

  if len(args) < 2:
    interface.reply_message('Please provide a recipient and message to send.')
    raise Exception('missing recipient or message')

  if not user.get('cookies'):
    interface.reply_message('Please log into TextNow first.')
    raise Exception('missing login cookies')

  interface.reply_message('Sending TextNow message...')

  try:
    recipient = args[0]
    message = ' '.join(args[1:])

    def bot(page):
      bot = textnow_bot.TextNowBot(page, user['cookies'])
      bot.send_message(recipient, message)

    playwright = Playwright(event['userId'])
    playwright.execute(bot)

    interface.reply_message('Successfully sent TextNow message!')
  except:
    screenshot_path = playwright.get_screenshot_path()

    if screenshot_path.exists():
      interface.reply_photo(screenshot_path)

    interface.reply_message('Failed to send TextNow message.')
    raise
